Constant propagation for individual fields through byte extract

tautschnig opened the pull request:
Field sensitivity in goto symex updates the full object as well as
individual fields for divisible objects. The updates of individual
fields, however, previously used index or member expressions on that
full object. Consequently, constant propagation was only possible when
the full-object assignment had a constant as the right-hand side. In
some cases, however, only parts of the whole object might be constant.
We are now able to constant propagate the fields covering those
(constant) parts.

Fixes: #6846

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Constant propagation for individual fields through byte extract

Field sensitivity in goto symex updates the full object as well as
individual fields for divisible objects. The updates of individual
fields, however, previously used index or member expressions on that
full object. Consequently, constant propagation was only possible when
the full-object assignment had a constant as the right-hand side. In
some cases, however, only parts of the whole object might be constant.
We are now able to constant propagate the fields covering those
(constant) parts.

Fixes: #6846

tautschnig committed:
Field sensitivity: do not simplify unconditionally

The option --no-simplify should be honoured by field sensitivity. This
also made apparent that we have tests that only pass thanks to the
simplifier, and perhaps aren't even expected to pass. See #6101 for
further discussion.

TGWDB reviewed (approved):
Generally happy to approve this on the basis that it improves the constant propagation and the "failing" tests are (as discussed in #6101) doing things that are not necessarily well defined.

tautschnig committed:
Increase timeout of incremental loop unwinding tests

8 seconds seemed arbitrary, and we have now seen a failure on one
platform, which might be just down to jitter. Increase the timeout to 15
seconds, which should be plenty.
